presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java

```diff
@@ -0,0 +1,99 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+import io.prestosql.plugin.kafka.encoder.RowEncoder;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ConnectorPageSink;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static io.prestosql.plugin.kafka.KafkaErrorCode.KAFKA_PRODUCER_ERROR;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class KafkaPageSink
+        implements ConnectorPageSink
+{
+    private final String topicName;
+    private final List<KafkaColumnHandle> columns;
+    private final RowEncoder keyEncoder;
+    private final RowEncoder messageEncoder;
+    private final KafkaProducer<byte[], byte[]> producer;
+
+    public KafkaPageSink(
+            String topicName,
+            List<KafkaColumnHandle> columns,
+            RowEncoder keyEncoder,
+            RowEncoder messageEncoder,
+            PlainTextKafkaProducerFactory producerFactory)
+    {
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.columns = requireNonNull(ImmutableList.copyOf(columns), "columns is null");
+        this.keyEncoder = requireNonNull(keyEncoder, "keyEncoder is null");
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+        requireNonNull(producerFactory, "producerFactory is null");
+        this.producer = producerFactory.create();
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            for (int channel = 0; channel < page.getChannelCount(); channel++) {
+                if (columns.get(channel).isKeyCodec()) {
+                    keyEncoder.appendColumnValue(page.getBlock(channel), position);
+                }
+                else {
+                    messageEncoder.appendColumnValue(page.getBlock(channel), position);
+                }
+            }
+            producer.send(new ProducerRecord<>(topicName, keyEncoder.toByteArray(), messageEncoder.toByteArray()), (recordMetadata, e) -> {
+                if (e != null) {
```

Comment:
The callback can be run in very different thread than the one calling `producer.send`. Do we know if exception thrown by the callback will be always propagated back to Presto thread. I.e if we do not get exception directly here, will we still get it on `producer.flush()` called from `finish()`?

Reply:
the callback is executed in the I/O thread of the producer. It doesn't look like flush throws an exceptions other than `InterruptException`

Reply:
Indeed it looks so. If the callback throws exception in later stage of sending it will be caught by try-catch in `ProducerBatch.completeFutureAndFireCallbacks` and just logged. See quoted method body below.
```java
    private void completeFutureAndFireCallbacks(long baseOffset, long logAppendTime, RuntimeException exception) {
        // Set the future before invoking the callbacks as we rely on its state for the `onCompletion` call
        produceFuture.set(baseOffset, logAppendTime, exception);

        // execute callbacks
        for (Thunk thunk : thunks) {
            try {
                if (exception == null) {
                    RecordMetadata metadata = thunk.future.value();
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(metadata, null);
                } else {
                    if (thunk.callback != null)
                        thunk.callback.onCompletion(null, exception);
                }
            } catch (Exception e) {
                log.error("Error executing user-provided callback on message for topic-partition '{}'", topicPartition, e);
            }
        }

        produceFuture.done();
    }
```

Given such logic I think that instead of throwing exception in from our callback, we should just record that message sending failure happened. A simple counter, how many message-sending failures we observed should do. 
Then in `KafkaPageSink.finish()`, after `producer.close()`, we can introspect the counter, and if it is >0 we can throw `PrestoException` there

Reply:
that sounds good!